File: CS333 - Programming Languages/File IO/Go/task1.go
/*
 * File: task1.go
 * Purpose: To create a program that reads a file and store the word frequency of each word in the file
 * Author: Samuel Munoz
 * Date: 04-14-2020
 */
package main
import (
	"fmt"
	"strings"
	"io/ioutil"
	"os"
)

type KeyValuePair struct {
	key string
	value int
}

func updateMap(m map[string]int, key string) {
	// if the state (represents if key given exists in the map) is true, then key exist and the value needs updating
	// otherwise, create new key with value one
	if _, state := m[key]; state {
		m[key] += 1
	} else {
		m[key] = 1
	}
}

func splitString(str string) []string {
	// splits string into the different lines
	a1 := strings.Split(str, "\n")
	a2 := make([]string, 0)
	a3 := make([]string, 0)
	a4 := make([]string, 0)
	for _, element := range a1 {
		temp := strings.Split(element, ". ")
		for _, s := range temp {
			a2 = append(a2, s)
		}
	}
	for _, element := range a2 {
		temp := strings.Split(element, ", ")
		for _, s := range temp {
			a3 = append(a3, s)
		}
	}
	for _, element := range a3 {
		temp := strings.Split(element, " ")
		for _, s := range temp {
			a4 = append(a4, s)
		}
	}
	return a4
}

func bubbleSort(array []*KeyValuePair) {
	length := len(array)
	for i := 0; i < length; i++ {
		for j := 0; j < length - i - 1; j++ {
			if array[j].value < array[j+1].value {
				temp := array[j]
				array[j] = array[j+1]
				array[j+1] = temp
			}
		}
	}
}

func main() {
	// stops the program if an argument is missing
	if(len(os.Args) < 2) {
		fmt.Printf("command: ./task1 <filename>\n")
		return
	}
	
	data, err := ioutil.ReadFile(os.Args[1]);
	// checks if there was an error in reading the file
	if err != nil {
		fmt.Printf("The program is unable to read the file.\n")
		return
	}

	// this map will hold all the 
	m := make(map[string]int)

	a := splitString(string(data))
	for _, element := range a {
		// update map with new keys
		updateMap(m, strings.ToLower(element))
	}

	array := make([]*KeyValuePair, 0)
	for key, value := range m {
		array = append(array, &KeyValuePair {
			key: key,
			value: value,
		})
	}

	bubbleSort(array)

	for i := 0; i < 20; i++ {
		fmt.Printf("%s\t\t\t%d\n", array[i].key, array[i].value)
	}
}

/*
Sources
https://blog.golang.org/maps
https://golang.org/pkg/strings/#Split
https://yourbasic.org/golang/convert-string-to-byte-slice/
https://tour.golang.org/moretypes/6
https://stackoverflow.com/questions/2050391/how-to-check-if-a-map-contains-a-key-in-go
https://gobyexample.com/reading-files
*/

File: CS333 - Programming Languages/File IO/C#/task1.cs
/*
 * File: Samuel Munoz
 * Purpose: To create a program that reads a file and prints the most frequency in a text file
 * Author: Samuel Munoz
 * Date: 04-13-2020
 */
using System;
using System.Collections.Generic; // package holds the dictionary class
using System.IO; // package holding the file object

namespace task1
{
    class Program
    {
    	static void updateTable(Dictionary<string, int> d, string key) {
    		// if the key already exist inside the hash table, then increase the frequency by one
    		if(d.ContainsKey(key)) {
    			d[key] = d[key] + 1;
    		}
    		// if the key does not exist, create a new key with value of one (frequency)
    		else {
    			d.Add(key, 1);
    		}
    	}

    	// sorts an array
    	static void bubbleSort(KeyValuePair<string, int>[] array) {
    		int length = array.Length;
    		for(int i=0; i < length-1; i++) {
    			for(int j=0; j < length-i-1; j++) {
    				int value1 = array[j].Value;
    				int value2 = array[j+1].Value;
    				if(value1 < value2) {
    					KeyValuePair<string, int> temp = array[j];
    					array[j] = array[j+1];
    					array[j+1] = temp;
    				}
    			}
    		}
    	}

        static void Main(string[] args)
        {
        	// if an argument is not given, throw error and exit program
        	if(args.Length < 1) {
        		Console.WriteLine("command: dotnet run <filename>");
        		return;
        	}
        	// if the file does not exist, throw error and exit the program
        	if(!File.Exists("./" + args[0])) {
        		Console.WriteLine("File does not exist. Please pass a file that exists");
        		return;
        	}

        	// this will store all the words as keys and their frequencies as values
            Dictionary<string, int> d = new Dictionary<string, int>();

         	// create read stream to read all the words in the file and store the words into the hash table
         	using(StreamReader sr = File.OpenText("./" + args[0])) {
         		string line;
         		while((line = sr.ReadLine()) != null) {
         				String[] words = System.Text.RegularExpressions.Regex.Split(line, @"[(\s)|(,\s)(.\s)]");
         				foreach(string word in words) {
         					// Console.WriteLine(word);
         					updateTable(d, word.ToLower());
         				}
         		}
         	}

         	KeyValuePair<string, int>[] entries = new KeyValuePair<string, int>[d.Count];
         	int index = 0;
         	foreach(KeyValuePair<string, int> kvp in d) {
         		entries[index] = kvp;
         		index++;
         	}

         	bubbleSort(entries);
         	for(index = 1; index < 21; index++)
         		Console.WriteLine("{0}\t\t\t{1}", entries[index].Key, entries[index].Value);
        }
    }
}

/*
Sources:
https://www.tutorialspoint.com/csharp/csharp_hashtable.htm
https://docs.microsoft.com/en-us/dotnet/api/system.collections.hashtable?view=netframework-4.8
https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/types/boxing-and-unboxing
https://docs.microsoft.com/en-us/dotnet/api/system.string.split?view=netframework-4.8
*/